Gradio 4

AUTOMATIC1111 opened the pull request:
Gradio 4 update.

The update breaks a lot of compatibility, some for very questionable reasons. Among other things, this branch makes it again possible to send bogus keyword arguments to components, and only get a warning for it, and use a `_js` argument for events. Basic functionality works, some extensions I have enabled like imagenet and tagcomplete load.

AUTOMATIC1111 committed:
gradio4

AUTOMATIC1111 committed:
remove the code that breaks extra networks

AUTOMATIC1111 committed:
make extra networks work again

AUTOMATIC1111 committed:
linter

missionfloyd committed:
Fix fullscreen image viewer

missionfloyd committed:
Fix popup CSS (mostly)

Center image buttons

missionfloyd committed:
Fix img2img

missionfloyd commented:
@daswer123 Basic img2img is working. Still need to fix sketch, inpaint, etc.

missionfloyd committed:
Fix inpaint

daswer123 commented:
Yeah, thanks, in general everything is more or less for build-in "canvas-zoom-and-pan", a number of features will have to be abandoned because it is no longer needed, some will have to be redone.

missionfloyd committed:
Change img2img_selected_tab back to gr.State

missionfloyd committed:
Fix sketch, inpaint sketch

Seems to work right, anyway.
Added webcam source.
Some img2img modes may now be redundant.

missionfloyd committed:
Remove unused import

missionfloyd committed:
Simplify inpaint sketch mask

missionfloyd committed:
Fix aspect ratio overlay

Make it work on inpaint upload tab

missionfloyd committed:
Fix "detect image size" button

AUTOMATIC1111 commented:
Oh, nice, there's been some work on this. Thanks. I couldn't do any myself because I was too busy.

AlUlkesh commented:
fwiw - as far as extensions go, while testing the Image Browser extension, I ran into ASGI errors, when many events were triggered. I addressed this with the Gradio team and is it fixed in Gradio 4.9.0.

w-e-w commented:
> fwiw - as far as extensions go, while testing the Image Browser extension, I ran into ASGI errors, when many events were triggered. I addressed this with the Gradio team and it is fixed in Gradio 4.9.0.

so bump to gradio 4.9.1?

w-e-w committed:
fix extras caption BLIP

missionfloyd committed:
Make extras work again

Not all postprocessing scripts work

missionfloyd committed:
Lint

missionfloyd committed:
Fix saving images from gallery

missionfloyd committed:
Remove unused import

missionfloyd committed:
Fix img2img interrogate

missionfloyd committed:
Fix extras BLIP caption

missionfloyd committed:
Expand gr.Image() dropzone to fill component

missionfloyd committed:
Merge branch 'gradio4' of https://github.com/automatic1111/stable-diffusion-webui into gradio4

missionfloyd committed:
Fix displaying images that haven't already been saved

Still copies already_saved_as images to temp.

AUTOMATIC1111 committed:
bump gradio version

AUTOMATIC1111 committed:
make it possible again to serve saved pictures without writing them to a temporary directory

AUTOMATIC1111 committed:
align the image in the gallery

AUTOMATIC1111 committed:
remove img2img editor height option, because it breaks gradio 4 image editors

AUTOMATIC1111 committed:
repair resolution calculation for img2img

AUTOMATIC1111 commented:
I added a bit more fixes. So far, copy image to <tab> buttons are broken - from txt2img to img2img, and from img2img to img2img. For different reasons.

AUTOMATIC1111 committed:
Merge branch 'dev' into gradio4

AUTOMATIC1111 committed:
remove gradio warning for refiner checkpoint

AUTOMATIC1111 committed:
solve some of issues with img2img copy to tab functionality

AUTOMATIC1111 committed:
linter

AUTOMATIC1111 committed:
fix js error at startup

AUTOMATIC1111 committed:
mark output gallery as non-editable

AUTOMATIC1111 committed:
fix send to tab and hires upscale button

AUTOMATIC1111 commented:
Fixed send to tabs buttons, updated to latest dev;
Current problems:
- inpaint tab is still misbehaving if you send an image to it from anywhere (could be just a gradio bug)
- bunch of errors in JS console at startup (could also be a gradio bug)

Mostly it feels like it works.

AUTOMATIC1111 committed:
lint

daswer123 committed:
Fix the startup zoom error

daswer123 committed:
fix fitToScreen and adjustBrushSize funcs in zoom.js

daswer123 committed:
Temporary fix that returns functionality when sending via buttons

daswer123 committed:
Made the zoom functionality work, both for drawing and erasing

daswer123 committed:
Remove unused code and lint

daswer123 committed:
place the cursor next to the original

daswer123 commented:
Hi all, made a PR where I tweaked the work of the built-in zoom extension #14816 

In general everything works, I study possible variants and in case I find bugs I will fix, also there are some thoughts on how to improve this extension.

In gradio 4 elements that work with the image have a lot of new buttons and functions, I thought it might be cool to realize convenient work with these functions through hotkeys.

For example, quick switching between drawing and erasing, quick layer creation, switching between layers via hotkeys. This is the first thing that came to my mind, I think there are many other things that can be realized.

I would like to know your opinion, is it worth working on the built-in extension or is it better to use these ideas to work on a separate extension?

Translated with www.DeepL.com/Translator (free version)

aolko commented:
Does ui still lag with some extensions installed on gradio 4?

AUTOMATIC1111 committed:
Merge pull request #14816 from daswer123/zoom-fix

Zoom & Pan: fix for gradio 4

daswer123 committed:
Fixed a bug with the cursor size when scrolling

When scrolling there is a bug, a gradio bug, because of which a parameter breaks the cursor size and it becomes bigger, so I made a solution that gets rid of this problem

daswer123 committed:
fix caused unnecessary borders when the cursor leaves the drawing area

daswer123 committed:
moved the part of code to a more suitable place, lint

AUTOMATIC1111 committed:
Merge pull request #14818 from daswer123/zoom-fix

Zoom & Pan: More fixes for gradio 4

w-e-w committed:
remove timestamp from path

AUTOMATIC1111 committed:
update to latest gradio

AUTOMATIC1111 committed:
Merge branch 'dev' into gradio4

AUTOMATIC1111 committed:
more fixes for latest gradio

AUTOMATIC1111 committed:
linter

AUTOMATIC1111 commented:
I think this can be merged now. Anyone care to test one more time?

AUTOMATIC1111 committed:
fix line endings for gradio.js

AUTOMATIC1111 committed:
Merge branch 'dev' into gradio4

AUTOMATIC1111 commented:
image editor is extremely unstable still in gradio

https://github.com/gradio-app/gradio/issues?q=is%3Aissue+is%3Aopen+ImageEditor

Basically, everything seems to work well except for copying images between painting tabs.

catboxanon committed:
Merge branch 'dev' into gradio4

catboxanon committed:
Fix merge errors

AUTOMATIC1111 commented:
Probably when they fix the bug with image inputs. Gradio 4 is a sad affair. It very likely will break extensions big time. Ooba updated to 4 and then reverted.

missionfloyd committed:
Bump gradio to 4.27.0

missionfloyd commented:
That seems to fix the send image buttons.

missionfloyd committed:
Fix empty override settings

Apparently empty dropdowns now return None

missionfloyd committed:
Lint

missionfloyd committed:
Update move_files_to_cache

missionfloyd committed:
Merge branch 'dev' into gradio4

missionfloyd committed:
Lint

w-e-w commented:
ya auto's issue post is closed
- https://github.com/gradio-app/gradio/issues/7591

missionfloyd committed:
Remove unused import

missionfloyd committed:
Fix save button

missionfloyd committed:
Make gallery return images

missionfloyd committed:
Get API working

Docs still doesn't work.

missionfloyd committed:
Change init_images default back to None

missionfloyd committed:
Copy img2img layers

missionfloyd committed:
Merge branch 'gradio4' of https://github.com/AUTOMATIC1111/stable-diffusion-webui into gradio4

missionfloyd committed:
Fix sending images from gallery

missionfloyd committed:
Fix img2img parameters

missionfloyd committed:
Fix extras task id

Fix restore extension config

huchenlei commented:
Is there any timeline when gradio 4.0 update will be ready?

missionfloyd committed:
Fix image editor layer menu appearance

huchenlei commented:
I am testing sd-webui-controlnet against gradio4 setup. However, a bunch of front-end issue seems to occur whenever I interact with almost anything on the extension UI. Any idea on this?
 
![image](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/20929282/72d554e3-96d8-4f56-9e72-3611c4b7b932)

AUTOMATIC1111 commented:
I don't know when I will merge this in. With SD3, it definitely should be after that.

I also want to get some info about which big extensions this breaks, and possibly fix compatibility issues from our side.

FurkanGozukara commented:
I bet Gradio 4 would break so many stuff

SD3 must be more priority I agree thank you so much @AUTOMATIC1111

Avaray commented:
Can't wait for [FileExplorer](https://www.gradio.app/docs/gradio/fileexplorer) 🤤

light-and-ray commented:
Forge canvas looks like a good think that can fix all inconvenience with gradio 4 image editor https://github.com/lllyasviel/stable-diffusion-webui-forge/discussions/853 I didn't test it, but porting it into our project can be a good idea

![352638054-5324f4f5-ad74-4342-a9e9-85fff01a3ede](https://github.com/user-attachments/assets/8146d883-ad55-4b05-ac98-c16a9e1e44ad)

missionfloyd committed:
Don't import Brush separately

missionfloyd committed:
Hide layers menu for inpaint

missionfloyd committed:
Merge branch 'dev' into gradio4

missionfloyd commented:
Gradio 5 breaks it all over again, because of course it does.
![image](https://github.com/user-attachments/assets/3ff9dbe1-af0f-486c-afec-8d5a3370f4d4)

missionfloyd committed:
Bump Gradio to 4.42.0

Fixes the API docs

light-and-ray commented:
Why they do it 😡

missionfloyd committed:
Downgrade Gradio to 4.38.1

ImageEditor color switching is broken on newer versions.

Qodo-Free-For-OSS commented:
Hi, Canvas zoom/pan assumes specific Gradio DOM nodes exist and immediately dereferences query results; if any selector fails (common under Gradio 4 markup/classname changes), it throws and breaks the extension.

**Severity:** action required | **Category:** reliability

**How to fix:** Add guards for missing nodes

**Agent prompt to fix** - you can give this to your LLM of choice:

> ### Issue description
> `zoom.js` assumes internal Gradio DOM structure; missing nodes cause runtime exceptions.
> 
> ### Issue Context
> Gradio 4 changes DOM structure and class names; selectors like `span.svelte-btgkrd` are brittle.
> 
> ### Fix Focus Areas
> - extensions-builtin/canvas-zoom-and-pan/javascript/zoom.js[289-383]
>   - Guard `eraseButton`/`paintButton` existence before adding listeners.
>   - Verify `canvasCursors.length >= 2` before indexing/cloning.
>   - Guard `canvasArea` and `canvasBorder` before dereferencing.
>   - Consider feature-detecting cursor elements instead of hardcoded hashed classes.

---
*Found by [Qodo](https://qodo.ai) code review*
